README
SIFTS

The SIFTS module is a Perl interface to the SIFTS mapping of PDB to 
UniProt residues. The SIFTS mapping is maintained at the EBI. For 
more information on the mapping itself, please refer to 
http://www.ebi.ac.uk/pdbe/docs/sifts/.

This module is designed for *nix systems exclusively. I have written 
a number of basic tests, but more is always best when it comes to 
tests. Please do not hesitate to contact me if you need more 
information or if you wish to contribute to this project.

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc SIFTS

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=SIFTS

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/SIFTS

    CPAN Ratings
        http://cpanratings.perl.org/d/SIFTS

    Search CPAN
        http://search.cpan.org/dist/SIFTS/


LICENSE AND COPYRIGHT

Copyright (C) 2011 Benoit H Dessailly

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for SIFTS

0.01    Date/time
        First version, released on an unsuspecting world.

0.02    2012-02-10/16:42:00 (+0900 GMT)
        Negative residue positions are now accepted in 
        SIFTS::Residue.
        
0.03    2012-02-23/11:17:00 (+0900 GMT)
        domid_to_pdbchains now warns when a chain from the 
        query file is not found in the SIFTS mapping file.

lib/SIFTS.pm
package SIFTS;

use 5.006;
use strict;
use warnings;

=head1 NAME

SIFTS is a Perl Module for parsing the PDB to UniProt residue 
mappings provided by project SIFTS at the EBI.

=head1 VERSION

Version 0.03

=cut

our $VERSION = '0.03';


=head1 SYNOPSIS

This module provides parsers to the different SIFTS mappings. 
Several scripts also come along this module for simple command line 
tasks. 

=head2 Scripts

=over 3

=item * B<domid_to_pdbchains>

Reads a list of CATH domain IDs from an input file and returns the 
list of PDB chains that come from the same protein chain as each of 
these domain IDs. The script is typically called as follow. Use 
option -h for more details on options.

  prompt# domid_to_pdbchains -i domids.dat -o domid_pdbchains.dat 

=back

=head2 Parsers

All parser classes are under SIFTS::IO.

=over 3

=item * B<SIFTS::IO::Text::PdbUniprot>

Parser for the text file mapping PDB chain IDs to Uniprot accession 
codes 'pdb_chains_uniprot.lst'. 

=back

=head2 Entity objects

The following modules are used when processing the mapping data 
files to store the information in a natural hierarchical 
object-oriented framework.

=over 3

=item * B<SIFTS::Dataset>

Stores a SIFTS dataset. A dataset corresponds to a set of at least
one SIFTS::Protein object. 

=item * B<SIFTS::Protein> 

Stores a protein. A protein corresponds to a Uniprot entry. A 
SIFTS::Protein objects can contain several SIFTS::Chain objects. 

=item * B<SIFTS::Chain>

Stores a protein chain. A protein chain corresponds to a PDB chain. 
A SIFTS::Chain object can contain several SIFTS::Segment objects.

=item * B<SIFTS::Segment>

Stores a protein chain segment. A protein chain segment corresponds 
to a continuous stretch of residues that can be mapped between the 
PDB chain and the Uniprot entry. A SIFTS::Segment object can contain
several SIFTS::Residue objects.

=item * B<SIFTS::Residue>

Stores a protein residue.   

=back

=head1 AUTHOR

Benoit H Dessailly, C<< <benoit at nibio.go.jp> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-sifts at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=SIFTS>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc SIFTS


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=SIFTS>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/SIFTS>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/SIFTS>

=item * Search CPAN

L<http://search.cpan.org/dist/SIFTS/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2011 Benoit H Dessailly.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


=cut

1; # End of SIFTS


lib/SIFTS/Chain.pm
package SIFTS::Chain;

use 5.006;
use strict;
use warnings;

use Carp;

=head1 NAME

SIFTS::Chain - Object representation of a protein chain in SIFTS.

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';


=head1 SYNOPSIS

Represents a protein chain in the context of the SIFTS mapping.

    use SIFTS::Chain;

    my $sifts_chain = SIFTS::Chain->new();
    
    ## Set Chain ID.
    $sifts_chain->pdbchainid( '1n3lA' );
    
    ## Assign a new SIFTS::Segment to the Chain.
    my $sifts_segment = SIFTS::Segment->new();
    $sifts_chain->add_segment( $sifts_segment );
    
    ## Set/Get a list of SIFTS::Segment objects that constitute the 
    ## Chain.
    my $sifts_segment1 = SIFTS::Segment->new();
    my $sifts_segment2 = SIFTS::Segment->new();
    my @sifts_segments = ( $sifts_segment1, $sifts_segment2 );
    $sifts_chain->segments( \@sifts_segments );
    
    ## Retrieve a SIFTS::Segment with a particular Segment ID from 
    ## the SIFTS::Chain object.
    my $sifts_segment = $sifts_chain->get_segment( 
        id => '1n3lA_1_173' 
    );

=head1 SUBROUTINES/METHODS

=head2 new

    my $sifts_chain = SIFTS::Chain->new();
    
  SIFTS::Chain::new creates a new empty SIFTS Protein Chain object.

=cut
sub new {
    my $class = shift;

    my $self = {};
    $self->{PDBCHAINID} = undef;
    $self->{SEGMENTS}   = {};

    bless( $self, $class );

    return $self;
}

=head2 pdbchainid

    $sifts_chain->pdbchainid( '1n3lA' );
    
  SIFTS::Chain::pdbchainid gets a PDB chain ID as argument for 
  assignment. The PDB chain ID should always consist of a PDB ID 
  followed by a one-letter chain ID. Always returns a PDB chain ID 
  or undef. 

=cut
sub pdbchainid {
    my $self = shift;
    my $val  = shift;
    
    if ( defined $val ) {

        ## Accepted format is a string of 5 alphanumeric characters.
        if ( $val =~ /^\w{5}$/ ) {
            $self->{PDBCHAINID} = $val;
        }
        else {
            carp "Warning: pdbchainid not assigned due to wrong ",
                 "format ($val).";
        }
    }
    
    return $self->{PDBCHAINID};
}

=head2 add_segment

    my $sifts_segment = SIFTS::Segment->new();
    $sifts_chain->add_segment( $sifts_segment );
    
  SIFTS::Chain::add_segment gets a L<SIFTS::Segment> compliant 
  object as argument for assignment. The L<SIFTS::Segment> object 
  will be added to the current list of Segments in the SIFTS Protein 
  Chain. The segment should have a defined id. Returns 1 upon 
  success. 

=cut
sub add_segment {
    my $self          = shift;
    my $sifts_segment = shift;
    
    if ( ! defined $sifts_segment ) {
        confess "Error: SIFTS::Chain->add_segment expects an ",
                "argument for assignment.";
    }

    if ( $sifts_segment->isa( 'SIFTS::Segment' ) != 1 ) {
        confess "Error: SIFTS::Chain->add_segment only takes ",
                "SIFTS::Segment compliant objects for assignment.";
    }
    
    if ( ! defined $sifts_segment->id ) {
        confess "Error: SIFTS::Chain->add_segment only takes a ",
                "SIFTS::Segment with defined id for assignment.";
    }

    $self->{SEGMENTS}{ $sifts_segment->id } = $sifts_segment;

    return 1;
}

=head2 segments

    ## Assign.
    my $sifts_segment1 = SIFTS::Segment->new();
    my $sifts_segment2 = SIFTS::Segment->new();
    my @sifts_segments = ( $sifts_segment1, $sifts_segment2 );
    $sifts_chain->segments( \@sifts_segments );
    
    ## Retrieve.
    my $sifts_segments_aref = $sifts_chain->segments();
    
  SIFTS::Chain::segments gets a reference to an array of 
  L<SIFTS::Segment> compliant objects as argument for assignment. 
  All segments in the array should have a defined id. Always returns 
  a reference to an array of L<SIFTS::Segment> objects contained in 
  the Protein Chain. The returned array may be empty.

=cut
sub segments {
    my $self                = shift;
    my $sifts_segments_aref = shift;
    
    if ( defined $sifts_segments_aref ) {
        
        for my $sifts_segment ( @{ $sifts_segments_aref } ) {
            
            ## Check that segment object is SIFTS::Segment compliant.
            if ( $sifts_segment->isa( 'SIFTS::Segment' ) != 1 ) {
                confess "Error: attempting to add a non ", 
                        "SIFTS::Segment compliant object into ",
                        "SIFTS::Chain->segments!";
            }
            
            ## Check that segment object has a defined id attribute.
            if ( defined $sifts_segment->id() ) {
                $self->{SEGMENTS}{ $sifts_segment->id() } 
                    = $sifts_segment;
            }
            else {
                confess "Error: attempting to add a SIFTS::Segment ",
                        "object with undefined id attribute into ",
                        "SIFTS::Chain->segments!";
            }
        }
    }
    
    my @segments = values %{ $self->{SEGMENTS} };
    
    return \@segments;
}

=head2 get_segment

    my $sifts_segment = $sifts_chain->get_segment( 
        id => '1n3lA_1_115',
    );
    
  SIFTS::Chain::get_segment gets a segment ID as argument and 
  searches the Protein Chain object for a L<SIFTS::Segment> object 
  with that ID. Returns the L<SIFTS::Segment> object if it finds it, 
  undef otherwise.

=cut
sub get_segment {
    my $self   = shift;
    my %arg    = @_;

    ## Check mandatory argument id is defined.
    confess "Missing mandatory argument id.\n"
        unless ( exists $arg{'id'} );

    my $segment = ( exists $self->{SEGMENTS}{$arg{'id'}} ) 
              ? $self->{SEGMENTS}{$arg{'id'}} : undef;
    
    return $segment;
}

=head1 AUTHOR

Benoit H Dessailly, C<< <benoit at nibio.go.jp> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-sifts-chain at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=SIFTS-Chain>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc SIFTS::Chain


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=SIFTS-Chain>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/SIFTS-Chain>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/SIFTS-Chain>

=item * Search CPAN

L<http://search.cpan.org/dist/SIFTS-Chain/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2011 Benoit H Dessailly.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


=cut

1; # End of SIFTS::Chain


lib/SIFTS/Dataset.pm
package SIFTS::Dataset;

use 5.006;
use strict;
use warnings;

use Carp;

=head1 NAME

SIFTS::Dataset - Object representation of a SIFTS mapping dataset.

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';


=head1 SYNOPSIS

Represents a SIFTS mapping dataset. 

    use SIFTS::Dataset;

    my $sifts_dataset = SIFTS::Dataset->new();
    
    ## Assign a new SIFTS::Protein object to the dataset.
    my $sifts_protein = SIFTS::Protein->new();
    $sifts_dataset->add_protein( $sifts_protein );
    
    ## Set a list of SIFTS::Protein objects that constitute the 
    ## dataset.
    my @sifts_proteins = ( $sifts_protein1, $sifts_protein_2 );
    $sifts_dataset->proteins( \@sifts_proteins );
    ## Get a list of SIFTS::Protein objects that constitute the 
    ## dataset.
    my $sifts_proteins_aref = $sifts_dataset->proteins();
    
    ## Retrieve a specific SIFTS::Protein object from the 
    ## SIFTS::Dataset.
    $sifts_dataset->get_protein( uniacc => 'P54399' )

=head1 SUBROUTINES/METHODS

=head2 new

    my $sifts_dataset = SIFTS::Dataset->new();
    
  SIFTS::Dataset::new creates a new empty SIFTS dataset object.

=cut
sub new {
    my $class = shift;

    my $self = {};
    $self->{PROTEINS} = {};

    bless( $self, $class );

    return $self;
}

=head2 add_protein

    $sifts_dataset->add_protein( $sifts_protein );
    
  SIFTS::Dataset::add_protein gets a L<SIFTS::Protein> compliant 
  object as argument for assignment. The protein should have a 
  defined Uniprot accession. The protein will be added to the 
  current list of Proteins in the SIFTS Dataset. Returns 1 upon
  success.

=cut
sub add_protein {
    my $self          = shift;
    my $sifts_protein = shift;
    
    if ( ! defined $sifts_protein ) {
        confess "Error: SIFTS::Dataset->add_protein expects a ",
                "SIFTS protein object as argument for assignment.";
    }
    
    if ( $sifts_protein->isa( 'SIFTS::Protein' ) != 1 ) {
        confess "Error: SIFTS::Dataset->add_protein expects a ",
                "SIFTS::Protein compliant object for assignment.";
    }
    
    if ( ! defined $sifts_protein->uniacc ) {
        confess "Error: SIFTS::Dataset->add_protein expects a ",
                "SIFTS::Protein object with defined uniacc for ",
                "assignment.";
    }
    
    $self->{PROTEINS}{ $sifts_protein->uniacc } = $sifts_protein;
    
    return 1;
}

=head2 proteins

    my @sifts_proteins = ( $sifts_protein1, $sifts_protein_2 );
    $sifts_dataset->proteins( \@sifts_proteins );
    
  SIFTS::Dataset::proteins gets a reference to an array of 
  L<SIFTS::Protein> compliant ojbects as argument for assignment.
  All proteins in the array reference should have defined uniprot
  accession attributes. Always returns a reference to an array of 
  Protein objects currently contained in the Dataset. The array may 
  be empty.  
  
=cut
sub proteins {
    my $self                = shift;
    my $sifts_proteins_aref = shift;
    
    if ( defined $sifts_proteins_aref ) {
        
        for my $sifts_protein ( @{ $sifts_proteins_aref } ) {
            
            ## Check that protein object is SIFTS::Protein compliant.
            if ( $sifts_protein->isa( 'SIFTS::Protein' ) != 1 ) {
                confess "Error: attempting to add a non ", 
                        "SIFTS::Protein compliant object into ",
                        "SIFTS::Dataset->proteins!";
            }
            
            ## Check that protein object has a defined uniacc 
            ## attribute.
            if ( defined $sifts_protein->uniacc() ) {
                $self->{PROTEINS}{ $sifts_protein->uniacc() } 
                    = $sifts_protein;
            }
            else {
                confess "Error: attempting to add a SIFTS::Protein ",
                        "object with undefined uniacc attribute ",
                        "into SIFTS::Dataset->proteins!";
            }
        }
    }
    
    my @proteins = values %{ $self->{PROTEINS} };
    
    return \@proteins;
}

=head2 get_protein

    my $sifts_protein = $sifts_dataset->get_protein( 
        uniacc => 'P54399', 
    );
    
  SIFTS::Dataset::get_protein gets a Uniprot accession code as 
  mandatory argument and searches the Dataset for a 
  L<SIFTS::Protein> object with that ID. Returns the 
  L<SIFTS::Protein> object if it finds it, undef otherwise.

=cut
sub get_protein {
    my $self   = shift;
    my %arg    = @_;

    ## Check mandatory argument uniacc is defined.
    confess "Missing mandatory argument uniacc.\n"
        unless ( exists $arg{'uniacc'} );

    my $protein = ( exists $self->{PROTEINS}{$arg{'uniacc'}} ) 
             ? $self->{PROTEINS}{$arg{'uniacc'}} : undef;
    
    return $protein;
}

=head1 AUTHOR

Benoit H Dessailly, C<< <benoit at nibio.go.jp> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-sifts-dataset at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=SIFTS-Dataset>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc SIFTS::Dataset


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=SIFTS-Dataset>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/SIFTS-Dataset>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/SIFTS-Dataset>

=item * Search CPAN

L<http://search.cpan.org/dist/SIFTS-Dataset/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2011 Benoit H Dessailly.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


=cut

1; # End of SIFTS::Dataset


lib/SIFTS/Protein.pm
package SIFTS::Protein;

use 5.006;
use strict;
use warnings;

use Carp;

=head1 NAME

SIFTS::Protein - Object representation of a protein in SIFTS.

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';


=head1 SYNOPSIS

Represents a protein in the context of the SIFTS mapping.

    use SIFTS::Protein;

    ## Create a new SIFTS Protein object.
    my $sifts_protein = SIFTS::Protein->new();
    
    ## Assign SIFTS Protein Uniprot Accession code.
    $sifts_protein->uniacc( 'P53432' );
    
    ## Assign a new SIFTS::Chain to the Protein.
    my $sifts_chain = SIFTS::Chain->new();
    $sifts_protein->add_chain( $sifts_chain );
    
    ## Set/Get a list of SIFTS::Chain objects that constitute the 
    ## protein.
    my $sifts_chain1 = SIFTS::Chain->new();
    my $sifts_chain2 = SIFTS::Chain->new();
    my @sifts_chains = ( $sifts_chain1, $sifts_chain2 );
    $sifts_protein->chains( \@sifts_chains );
    
    ## Retrieve a SIFTS::Chain with a particular Chain ID from the 
    ## SIFTS::Protein object.
    my $sifts_chain = $sifts_protein->get_chain( 
        pdbchainid => '1n3lA' 
    );

=head1 SUBROUTINES/METHODS

=head2 new

    my $sifts_protein = SIFTS::Protein->new();
    
  SIFTS::Protein::new creates a new empty SIFTS Protein object.

=cut
sub new {
    my $class = shift;

    my $self = {};
    $self->{UNIACC} = undef;
    $self->{CHAINS} = {};

    bless( $self, $class );

    return $self;
}

=head2 uniacc

    $sifts_protein->uniacc( 'P53093' );
    
  SIFTS::Protein::id gets a Protein Uniprot Accession code as 
  argument for assignment. The Protein Uniprot Accession code should
  always consist of a string with 6 alphanumeric characters. Always 
  returns the uniprot accession code or undef.  

=cut
sub uniacc {
    my $self = shift;
    my $val  = shift;
    
    if ( defined $val ) {

        ## Accepted format is a string of 6 alphanumeric characters.
        if ( $val =~ /^\w{6}$/ ) {
            $self->{UNIACC} = $val;
        }
        else {
            carp "Warning: id not assigned due to wrong format ",
                 "($val).";
        }
    }
    
    return $self->{UNIACC};
}

=head2 add_chain

    my $sifts_chain = SIFTS::Chain->new();
    $sifts_protein->add_chain( $sifts_chain );
    
  SIFTS::Protein::add_chain gets a L<SIFTS::Chain> compliant object 
  as argument for assignment. The L<SIFTS::Chain> object will be 
  added to the current list of Chains in the SIFTS Protein. The 
  chain should have a defined pdb chain ID. Returns 1 upon success. 

=cut
sub add_chain {
    my $self        = shift;
    my $sifts_chain = shift;
    
    if ( ! defined $sifts_chain ) {
        confess "Error: SIFTS::Protein->add_chain expects an ",
                "argument for assignment.";
    }

    if ( $sifts_chain->isa( 'SIFTS::Chain' ) != 1 ) {
        confess "Error: SIFTS::Protein->add_chain only takes ",
                "SIFTS::Chain compliant objects for assignment.";
    }
    
    if ( ! defined $sifts_chain->pdbchainid ) {
        confess "Error: SIFTS::Protein->add_chain only takes a ",
                "SIFTS::Chain with defined PDB chain ID for ",
                "assignment.";
    }

    $self->{CHAINS}{ $sifts_chain->pdbchainid } = $sifts_chain;

    return 1;
}

=head2 chains

    ## Assign.
    my $sifts_chain1 = SIFTS::Chain->new();
    my $sifts_chain2 = SIFTS::Chain->new();
    my @sifts_chains = ( $sifts_chain1, $sifts_chain2 );
    $sifts_protein->chains( \@sifts_chains );
    
    ## Retrieve.
    my $sifts_chains_aref = $sifts_protein->chains();
    
  SIFTS::Protein::chains gets a reference to an array of 
  L<SIFTS::Chain> compliant objects as argument for assignment. 
  All chains in the array should have a defined PDB chain ID.
  Always returns a reference to an array of L<SIFTS::Chain> objects 
  contained in the Protein. The returned array may be empty.

=cut
sub chains {
    my $self              = shift;
    my $sifts_chains_aref = shift;
    
    if ( defined $sifts_chains_aref ) {
        
        for my $sifts_chain ( @{ $sifts_chains_aref } ) {
            
            ## Check that chain object is SIFTS::Chain compliant.
            if ( $sifts_chain->isa( 'SIFTS::Chain' ) != 1 ) {
                confess "Error: attempting to add a non ", 
                        "SIFTS::Chain compliant object into ",
                        "SIFTS::Protein->chains!";
            }
            
            ## Check that chain object has a defined pdbchainid 
            ## attribute.
            if ( defined $sifts_chain->pdbchainid() ) {
                $self->{CHAINS}{ $sifts_chain->pdbchainid() } 
                    = $sifts_chain;
            }
            else {
                confess "Error: attempting to add a SIFTS::Chain ",
                        "object with undefined pdbchainid ",
                        "attribute into SIFTS::Protein->chains!";
            }
        }
    }
    
    my @chains = values %{ $self->{CHAINS} };
    
    return \@chains;
}

=head2 get_chain

    my $sifts_chain = $sifts_protein->get_chain( 
        pdbchainid => '1n3lA',
    );
    
  SIFTS::Protein::get_chain gets a PDB chain ID as argument and
  searches the Protein object for a L<SIFTS::Chain> object with that
  ID. Returns the L<SIFTS::Chain> object if it finds it, undef 
  otherwise.

=cut
sub get_chain {
    my $self   = shift;
    my %arg    = @_;

    ## Check mandatory argument pdbchainid is defined.
    confess "Missing mandatory argument pdbchainid.\n"
        unless ( exists $arg{'pdbchainid'} );

    my $chain = ( exists $self->{CHAINS}{$arg{'pdbchainid'}} ) 
              ? $self->{CHAINS}{$arg{'pdbchainid'}} : undef;
    
    return $chain;
}

=head1 AUTHOR

Benoit H Dessailly, C<< <benoit at nibio.go.jp> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-sifts-protein at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=SIFTS-Protein>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc SIFTS::Protein


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=SIFTS-Protein>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/SIFTS-Protein>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/SIFTS-Protein>

=item * Search CPAN

L<http://search.cpan.org/dist/SIFTS-Protein/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2011 Benoit H Dessailly.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


=cut

1; # End of SIFTS::Protein


lib/SIFTS/Segment.pm
package SIFTS::Segment;

use 5.006;
use strict;
use warnings;

use Carp;

=head1 NAME

SIFTS::Segment - Object representation of a protein chain segment in 
SIFTS.

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';


=head1 SYNOPSIS

Represents a Protein Chain Segment in the context of the SIFTS 
mapping.

    use SIFTS::Segment;

    ## Create a new Segment object.
    my $sifts_segment = SIFTS::Segment->new();
    
    ## Assign Segment ID.
    $sifts_segment->id( '1n3lA_1_115' );
    
    ## Set/Get Start Residue.
    $sifts_segment->start_residue( $sifts_residue );
    
    ## Set/Get Stop Residue.
    $sifts_segment->stop_residue( $sifts_residue );
    
=head1 SUBROUTINES/METHODS

=head2 new

    my $sifts_segment = SIFTS::Segment->new();
    
  SIFTS::Segment::new creates a new empty SIFTS Protein Chain 
  Segment object.

=cut
sub new {
    my $class = shift;

    my $self           = {};
    $self->{ID}        = undef;
    $self->{START_RES} = undef;
    $self->{STOP_RES}  = undef;

    bless( $self, $class );

    return $self;
}

=head2 id

    $sifts_segment->id( '1n3lA_1_115' );
    
  SIFTS::Segment::id gets a Segment ID as argument for assignment.
  The segment ID should always consist of a PDB ID, a Chain ID, 
  a '_' character, a start position, a '_' character and a stop
  position. Always returns the Segment ID or undef. 

=cut
sub id {
    my $self = shift;
    my $val  = shift;
    
    if ( defined $val ) {

        ## Accepted format is a string of 5 alphanumeric characters,
        ## followed by '_', followed by alphanumeric characters, 
        ## followed by '_', followed by alphanumeric characters.
        if ( $val =~ /^\w{5}_\w+_\w+$/ ) {
            $self->{ID} = $val;
        }
        else {
            carp "Warning: id not assigned due to wrong format ",
                 "($val).";
        }
    }
    
    return $self->{ID};
}

=head2 start_residue

    ## Set.
    $sifts_segment->start_residue( $sifts_residue );
    
    ## Get.
    my $start_residue = $sifts_segment->start_residue();
    
  SIFTS::Segment::start_residue gets a L<SIFTS::Residue> compliant
  object as argument for assignment. Always returns a 
  L<SIFTS::Residue> compliant object or undef.

=cut
sub start_residue {
    my $self      = shift;
    my $start_res = shift;
    
    if ( defined $start_res ) {
    
        if ( $start_res->isa( 'SIFTS::Residue') != 1 ) {
            confess "Error: SIFTS::Segment->start_residue only ",
                    "takes SIFTS::Residue compliant objects for ",
                    "assignment.";
        }
    
        $self->{START_RES} = $start_res;
    }
    
    return $self->{START_RES};
}

=head2 stop_residue

    ## Set.
    $sifts_segment->stop_residue( $sifts_residue );
    
    ## Get.
    my $stop_residue = $sifts_segment->stop_residue();
    
  SIFTS::Segment::stop_residue gets a L<SIFTS::Residue> compliant
  object as argument for assignment. Always returns a 
  L<SIFTS::Residue> compliant object or undef.

=cut
sub stop_residue {
    my $self     = shift;
    my $stop_res = shift;

    if ( defined $stop_res ) {

        if ( $stop_res->isa( 'SIFTS::Residue') != 1 ) {
            confess "Error: SIFTS::Segment->stop_residue only ",
                    "takes SIFTS::Residue compliant objects for ",
                    "assignment.";
        }

        $self->{STOP_RES} = $stop_res;
    }

    return $self->{STOP_RES};
}

=head1 AUTHOR

Benoit H Dessailly, C<< <benoit at nibio.go.jp> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-sifts-segment at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=SIFTS-Segment>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc SIFTS::Segment


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=SIFTS-Segment>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/SIFTS-Segment>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/SIFTS-Segment>

=item * Search CPAN

L<http://search.cpan.org/dist/SIFTS-Segment/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2011 Benoit H Dessailly.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


=cut

1; # End of SIFTS::Segment
